tiledb/sm/subarray/cell_slab_iter.cc

```diff
@@ -0,0 +1,324 @@
+    return Status::Ok();
+
+  RETURN_NOT_OK(sanity_check());
+  cell_slab_.init(subarray_->dim_num());
+  RETURN_NOT_OK(init_ranges());
+  init_coords();
+  init_cell_slab_lengths();
+  update_cell_slab();
+
+  end_ = false;
+
+  return Status::Ok();
+}
+
+template <class T>
+typename CellSlabIter<T>::CellSlab CellSlabIter<T>::cell_slab() const {
+  return cell_slab_;
+}
+
+template <class T>
+bool CellSlabIter<T>::end() const {
+  return end_;
+}
+
+template <class T>
+void CellSlabIter<T>::operator++() {
+  // If at the end, do nothing
+  if (end_)
+    return;
+
+  // Advance the iterator
+  if (subarray_->layout() == Layout::ROW_MAJOR)
+    advance_row();
+  else
+    advance_col();
+
+  if (end_) {
+    cell_slab_.reset();
+    return;
+  }
+
+  update_cell_slab();
+}
+
+/* ****************************** */
+/*          PRIVATE METHODS       */
+/* ****************************** */
+
+template <class T>
+void CellSlabIter<T>::advance_col() {
+  auto dim_num = (int)subarray_->dim_num();
+
+  for (int i = 0; i < dim_num; ++i) {
+    cell_slab_coords_[i] += (i == 0) ? cell_slab_lengths_[range_coords_[i]] : 1;
+    if (cell_slab_coords_[i] > ranges_[i][range_coords_[i]].end_) {
+      ++range_coords_[i];
+      cell_slab_coords_[i] = ranges_[i][range_coords_[i]].start_;
+    }
+
+    if (range_coords_[i] < (T)ranges_[i].size()) {
+      break;
+    } else {
+      // The iterator has reached the end
+      if (i == dim_num - 1) {
+        end_ = true;
+        return;
+      }
+
+      range_coords_[i] = 0;
+      cell_slab_coords_[i] = ranges_[i][0].start_;
+    }
+  }
+}
+
+template <class T>
+void CellSlabIter<T>::advance_row() {
+  auto dim_num = (int)subarray_->dim_num();
+
+  for (int i = dim_num - 1; i >= 0; --i) {
+    cell_slab_coords_[i] +=
+        (i == dim_num - 1) ? cell_slab_lengths_[range_coords_[i]] : 1;
+    if (cell_slab_coords_[i] > ranges_[i][range_coords_[i]].end_) {
+      ++range_coords_[i];
+      cell_slab_coords_[i] = ranges_[i][range_coords_[i]].start_;
+    }
+
+    if (range_coords_[i] < (T)ranges_[i].size()) {
+      break;
+    } else {
+      // The iterator has reached the end
+      if (i == 0) {
+        end_ = true;
+        return;
+      }
+
+      range_coords_[i] = 0;
+      cell_slab_coords_[i] = ranges_[i][0].start_;
+    }
+  }
+}
+
+template <class T>
+void CellSlabIter<T>::create_ranges(
+    const T* range,
+    T tile_extent,
+    T dim_domain_start,
+    std::vector<Range>* ranges) {
+  T tile_start = (range[0] - dim_domain_start) / tile_extent;
+  T tile_end = (range[1] - dim_domain_start) / tile_extent;
+
+  // The range falls int he same tile
+  if (tile_start == tile_end) {
+    ranges->emplace_back(range[0], range[1], tile_start);
+  } else {  // We need to split the range
+    T start = range[0];
+    T end;
+    for (T i = tile_start; i < tile_end; ++i) {
+      end = (i + 1) * tile_extent + dim_domain_start - 1;
+      ranges->emplace_back(start, end, i);
+      start = end + 1;
```

Comment:
Same question here re: `nextafter`

Reply:
`CellSlabIter` works only with dense domains, so `+1` here is safe.